File: graph/building_road.py
import collections
def main():
    n , m = map(int , input().split())
    adj_list = collections.defaultdict(list)
    for _ in range(m):
        a , b = map(int , input().split())
        adj_list[a].append(b)
        adj_list[b].append(a)

    visited = [False]*(n+1)
    def dfs(i):
        visited[i]=True
        for v in adj_list[i]:
            if not visited[v]:
                dfs(v)

    new_road = []
    for i in range(1,n+1):
        if not visited[i]:
            new_road.append(i)
            dfs(i)

    print(len(new_road)-1)
    for i in range(len(new_road)-1):
        print("{} {}".format(new_road[i],new_road[i+1]))



if __name__ == "__main__":
    main()

File: graph/building_road.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

const int MAX=1e5+10;
vector<int> adj_list[MAX], visited(MAX,0);

void dfs(int i){
    visited[i] = 1;
    for(auto &i: adj_list[i]){
        if (!visited[i]){
             dfs(i);
        }     
    }
}
int main()
{
    int n,m;
    cin>>n>>m;

    for (int i=0;i<m;i++){
        int u,v;
        cin>>u>>v;
        adj_list[u].push_back(v);
        adj_list[v].push_back(u);
    }
    vector<int>ans;
    for(int i =1;i<=n;i++){
        if (!visited[i]){
            ans.push_back(i);
            dfs(i);
        }
    }
    
    cout<< ans.size()-1 <<"\n";
    for(int i =0;i<ans.size()-1;i++){
        cout << ans[i] << " " <<ans[i+1] <<"\n";
    }
}

